src/preact-base-element.js

```diff
@@ -0,0 +1,314 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Deferred} from './utils/promise';
+import {Fragment, createElement, render} from 'preact';
+import {Slot, createSlot} from './preact/slot';
+import {devAssert} from './log';
+import {matches} from './dom';
+import {withAmpContext} from './preact/context';
+
+/**
+ * @typedef {{
+ *   attr: string,
+ *   type: (string|undefined),
+ *   default: *,
+ * }}
+ */
+export let AmpElementProp;
+
+/**
+ * @typedef {{
+ *   className: (string|undefined),
+ *   props: (!Object<string, !AmpElementProp>|undefined),
+ *   isLayoutSupported: (
+ *     function(
+ *       this:./base-element.BaseElement,
+ *       !./layout.Layout
+ *     ):boolean | undefined
+ *   ),
+ *   init: (
+ *     function(
+ *       this:./base-element.BaseElement
+ *     ):(!JsonObject|undefined) | undefined
+ *   ),
+ * }}
+ */
+export let AmpElementOptions;
+
+/**
+ * Wraps a Preact Component in a BaseElement class.
+ *
+ * Most functionality should be done in Preact. We don't expose the BaseElement
+ * subclass on purpose, you're not meant to do work in the subclass! There will
+ * be very few exceptions, which is why we allow options to configure the
+ * class.
+ *
+ * @param {function(!JsonObject):*} Component
+ * @param {!AmpElementOptions} opts
+ * @return {function(new:./base-element.BaseElement, !Element)}
+ */
+export function PreactBaseElement(Component, opts = {}) {
+  return class extends AMP.BaseElement {
+    /** @param {!AmpElement} element */
+    constructor(element) {
+      super(element);
+
+      /** @private {?Node} */
+      this.container_ = null;
+
+      /** @private {boolean} */
+      this.scheduledRender_ = false;
+
+      /** @private {!Object} */
+      this.context_ = {
+        renderable: false,
+        playable: false,
+        notify: () => this.mutateElement(() => {}),
+      };
+
+      this.boundRerender_ = () => {
+        this.scheduledRender_ = false;
+        this.rerender_();
+      };
+
+      /** @private {!Deferred|null} */
+      this.scheduledRenderDeferred_ = null;
+
+      /** @private {!JsonObject|null|undefined} */
+      this.defaultProps_ = null;
+
+      /** @private {boolean} */
+      this.mounted_ = true;
+    }
+
+    /** @override */
+    renderOutsideViewport() {
+      const distance = opts.renderOutsideViewport;
+      return distance == null ? super.renderOutsideViewport() : distance;
+    }
+
+    /** @override */
+    isLayoutSupported(layout) {
+      const layoutSupported = opts.isLayoutSupported;
+      return layoutSupported ? layoutSupported.call(this, layout) : true;
```

Comment:
Hmm. It's ok to delay this to followup PRs, but let's touch base one why this shouldn't just be a simple class extends way. Is it only because of props?

Reply:
I just haven't converted it yet. We can immediately change it in a follow up.